tests/sentry/event_manager/grouping/test_assign_to_group.py

```diff
@@ -0,0 +1,451 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from collections.abc import Callable
+from contextlib import contextmanager
+from time import time
+from typing import Any, ParamSpec, TypeVar
+from unittest import mock
+
+import pytest
+
+from sentry.event_manager import _create_group
+from sentry.eventstore.models import Event
+from sentry.grouping.ingest import (
+    _calculate_primary_hash,
+    _calculate_secondary_hash,
+    find_existing_grouphash,
+)
+from sentry.models.grouphash import GroupHash
+from sentry.models.project import Project
+from sentry.testutils.helpers.eventprocessing import save_new_event
+from sentry.testutils.pytest.fixtures import django_db_all
+from sentry.testutils.skips import requires_snuba
+
+pytestmark = [requires_snuba]
+
+
+LEGACY_CONFIG = "legacy:2019-03-12"
+NEWSTYLE_CONFIG = "newstyle:2023-01-11"
+
+# TODO: Once we're on python 3.12, we can get rid of these and change the first line of the
+# signature of `make_wrapped_function_with_captured_return_values` to
+#   def make_wrapped_function_with_captured_return_values[T, **P](
+P = ParamSpec("P")
+T = TypeVar("T")
+
+
+def make_wrapped_function_with_captured_return_values(
+    fn: Callable[P, T],
+    return_values: list[T],
+) -> Callable[P, T]:
+    """
+    Create a wrapped version of the given function, which stores the return value each time that
+    function is called. This is useful when you want to spy on a given function and make assertions
+    based on what it returns.
+
+    In a test, this can be used in concert with a patching context manager like so:
+
+        from unittest import mock
+        from wherever import make_wrapped_function_with_captured_return_values
+        from some_module import some_function
+
+        def test_doing_the_thing():
+            some_function_return_values = []
+            wrapped_some_function = make_wrapped_function_with_captured_return_values(
+                some_function, some_function_return_values
+            )
+
+            with mock.patch(
+                "some_module.some_function", wraps=wrapped_some_function
+            ) as some_function_spy:
+                a_function_that_calls_some_function()
+                assert some_function_spy.call_count == 1
+                assert some_function_return_values[0] == "some value"
+
+    """
+
+    def wrapped_fn(*args: P.args, **kwargs: P.kwargs) -> T:
+        returned_value = fn(*args, **kwargs)
+        return_values.append(returned_value)
```

Comment:
```suggestion
        return_values[fn.__name__].append(returned_value)
```

and then simplify the calling side like

```make_wrapped_function_with_captured_return_values(find_existing_grouphash, return_values)```

Reply:
Oh, nice. I wondered if there weren't a nicer way to do that, but then got distracted by... something. A squirrel? Who knows. 

In any case, good suggestion, thanks!

Reply:
Oh, wait, no. That'd work in this case, where I'm using the wrapping function multiple times and wanting to store all the results lists in a single dictionary, but in general, the idea is that you ought to be able to use it once and just pass it an empty list.

[UPDATE] I decided that it's useful both ways, so now they both work.